broker/src/main/java/io/zeebe/broker/system/partitions/impl/StateControllerImpl.java

```diff
@@ -0,0 +1,312 @@
+import io.atomix.raft.snapshot.PersistedSnapshotStore;
+import io.atomix.raft.snapshot.ReceivedSnapshot;
+import io.atomix.raft.snapshot.SnapshotChunk;
+import io.atomix.raft.snapshot.TransientSnapshot;
+import io.atomix.utils.time.WallClockTimestamp;
+import io.zeebe.broker.system.partitions.AtomixRecordEntrySupplier;
+import io.zeebe.broker.system.partitions.SnapshotReplication;
+import io.zeebe.broker.system.partitions.StateController;
+import io.zeebe.db.ZeebeDb;
+import io.zeebe.db.ZeebeDbFactory;
+import io.zeebe.logstreams.impl.Loggers;
+import io.zeebe.util.FileUtil;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.ToLongFunction;
+import org.agrona.collections.Object2NullableObjectHashMap;
+import org.slf4j.Logger;
+
+/** Controls how snapshot/recovery operations are performed */
+public class StateControllerImpl implements StateController, PersistedSnapshotListener {
+
+  private static final ReplicationContext INVALID_SNAPSHOT = new ReplicationContext(-1, null);
+  private static final Logger LOG = Loggers.SNAPSHOT_LOGGER;
+
+  private final SnapshotReplication replication;
+  private final Map<String, ReplicationContext> receivedSnapshots =
+      new Object2NullableObjectHashMap<>();
+
+  private final PersistedSnapshotStore store;
+
+  private final Path runtimeDirectory;
+  private final ZeebeDbFactory zeebeDbFactory;
+  private final ToLongFunction<ZeebeDb> exporterPositionSupplier;
+  private final AtomixRecordEntrySupplier entrySupplier;
+
+  private final SnapshotReplicationMetrics metrics;
+
+  private ZeebeDb db;
+
+  public StateControllerImpl(
+      final int partitionId,
+      final ZeebeDbFactory zeebeDbFactory,
+      final PersistedSnapshotStore store,
+      final Path runtimeDirectory,
+      final SnapshotReplication replication,
+      final AtomixRecordEntrySupplier entrySupplier,
+      final ToLongFunction<ZeebeDb> exporterPositionSupplier) {
+    this.store = store;
+    this.runtimeDirectory = runtimeDirectory;
+    this.zeebeDbFactory = zeebeDbFactory;
+    this.exporterPositionSupplier = exporterPositionSupplier;
+    this.entrySupplier = entrySupplier;
+    this.replication = replication;
+    this.metrics = new SnapshotReplicationMetrics(Integer.toString(partitionId));
+    store.addSnapshotListener(this);
+  }
+
+  @Override
+  public Optional<TransientSnapshot> takeTransientSnapshot(final long lowerBoundSnapshotPosition) {
+    if (!isDbOpened()) {
+      return Optional.empty();
+    }
+
+    final long exportedPosition = exporterPositionSupplier.applyAsLong(openDb());
+    final long snapshotPosition = Math.min(exportedPosition, lowerBoundSnapshotPosition);
+
+    final var optionalIndexed = entrySupplier.getIndexedEntry(snapshotPosition);
+
+    final Long previousSnapshotIndex =
+        store.getLatestSnapshot().map(PersistedSnapshot::getCompactionBound).orElse(-1L);
```

Comment:
This is actually from here https://github.com/zeebe-io/zeebe/blob/develop/atomix/cluster/src/main/java/io/atomix/raft/impl/zeebe/snapshot/AtomixSnapshotStorage.java#L59 and I think because it is the compaction bound you compare here but yes in the end it is the index